modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/taglib/JSLoaderTopHeadDynamicInclude.java

```diff
@@ -116,6 +131,8 @@ public void register(DynamicIncludeRegistry dynamicIncludeRegistry) {
 	protected void activate(Map<String, Object> properties) {
 		_details = ConfigurableUtil.createConfigurable(
 			Details.class, properties);
+
+		_lastModified = System.currentTimeMillis();
```

Comment:
:thinking: I'm not sure if this is going to work. The only place where you seem to update `_lastModified` is here, which means you only invalidate the cached content when the OSGi component is activated (i.e.: when the portal is started, after a redeployment of the module, or when the configuration changes, basically).

However, the cached content is composed with values that are not coming from the configuration (see [here](https://github.com/liferay-frontend/liferay-portal/pull/686/files#diff-0e8b09acc631a4e126cd033a1fc7c6f0af055af07b5ab80be88ad941dd3e3bccR86) and [here](https://github.com/liferay-frontend/liferay-portal/pull/686/files#diff-0e8b09acc631a4e126cd033a1fc7c6f0af055af07b5ab80be88ad941dd3e3bccR100) for example).

Won't this cause a problem?

Reply:
Specifically, [_getDefaultURLParamsJSONObject](https://github.com/liferay-frontend/liferay-portal/pull/686/files#diff-0e8b09acc631a4e126cd033a1fc7c6f0af055af07b5ab80be88ad941dd3e3bccR145) seems to depend on the specific ongoing request.

Reply:
Mmm... the links seem off, but yeah, I guess you mean everything depending on `themeDisplay`... that should indeed not be cached, I guess...

Thoughts on this? 🤔

Reply:
I also wonder how it worked before since it also uses `AbsolutePortalURLBuilder` and `MinifierUtil` which are environment sensitive, but maybe not so much... 🤔

Reply:
Well, we could simply cache that and print the rest always 🤷

Reply:
But if we cache that and print the rest I suppose the regression will stay present :grimacing: .

Regarding the `AbsolutePortalURLBuilder` and `MinifierUtil`, it's true. But since they can only change by changing the `portal.properties` file or redeploying (and AFAIK we don't support redeployment in production) it has never given any problem, as you pointed out.

Reply:
Ok, so we have 2 options that I can think of:
- Simply print the JS code we had before in Java, so that the values for `combine` and `languageId` are dynamic as they were before
- Have a multi-key cache, so we keep track of `combine` + `languageId` and access those objects if `_lastModified` is valid.

I quickly tried the first approach and it looks awful and sketchy, so I'd rather go with a more robust and stable caching, but I wouldn't mind going with the first approach if it gets to messy.